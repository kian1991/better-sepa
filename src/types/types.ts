import { z } from 'zod';
import { CreditTransferPaymentSchema, CreditTransferSchema, IBANSchema } from './schemas';

// inferred types
export type CreditTransferPayment = z.infer<typeof CreditTransferPaymentSchema>;
export type CreditTransferPaymentInput = z.input<typeof CreditTransferPaymentSchema>;
export type CreditTransferInput = z.input<typeof CreditTransferSchema>;
export type CreditTransfer = z.infer<typeof CreditTransferSchema>;
export type IBAN = z.infer<typeof IBANSchema>;
