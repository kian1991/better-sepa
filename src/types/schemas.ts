import { z } from 'zod';
import { validateIBAN } from '../lib/validation';
import { createMsgId, generateEndToEndReference } from '../lib/util';

// general schemas
export const IBANSchema = z.string().min(15).max(34).refine(validateIBAN, {
	message: 'Invalid IBAN',
});

// PAIN.001.001.03 schemas (credit transfer)
export const CreditTransferPaymentSchema = z.object({
	recipientName: z.string(),
	recipientIBAN: IBANSchema,
	recipientBIC: z.string().optional(),
	amount: z.number().positive(),
	reference: z.string().optional(),
	end2endReference: z.string().default(() => generateEndToEndReference()),
});

export const CreditTransferSchema = z.object({
	documentId: z.string().default(createMsgId(new Date())),
	documentCreationDate: z.date().default(() => new Date()),
	requestedExecutionDate: z.date().default(() => new Date()),
	initiatorName: z.string(),
	initiatorIBAN: IBANSchema,
	initiatorBIC: z.string(),
	batchBooking: z.boolean().default(false),
	payments: z.array(CreditTransferPaymentSchema),
});
