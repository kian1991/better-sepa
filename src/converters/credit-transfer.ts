import { create } from 'xmlbuilder2';
import {
  CreditTransfer,
  CreditTransferInput,
  CreditTransferPayment,
} from '../types/types';
import { CreditTransferSchema } from '../types/schemas';

export const createCreditTransferXML = (
  creditTransfer: CreditTransferInput
): [string, CreditTransfer] => {
  // Try to parse the input data and throw an error if it is invalid
  const parsingResult = CreditTransferSchema.safeParse(creditTransfer);

  if (!parsingResult.success) {
    throw new Error(parsingResult.error.toString());
  }

  const parsedCreditTransfer = parsingResult.data;
  console.log(parsedCreditTransfer);

  // Map the payments to the XML structure
  const mappedPayments = parsedCreditTransfer.payments.map((payment) => {
    // BIC Handling: check if BIC is provided. If not add the _Othr_ Attribute
    const parsedBIC = payment.recipientBIC
      ? { BIC: payment.recipientBIC }
      : {
          Othr: {
            Id: 'NOTPROVIDED',
          },
        };

    return {
      PmtId: {
        EndToEndId: payment.end2endReference,
      },
      Amt: {
        InstdAmt: {
          '@Ccy': 'EUR',
          '#text': payment.amount,
        },
      },
      CdtrAgt: {
        FinInstnId: {
          parsedBIC,
        },
      },
      Cdtr: {
        Nm: payment.recipientName,
      },
      CdtrAcct: {
        Id: {
          IBAN: payment.recipientIBAN,
        },
      },
      RmtInf: {
        Ustrd: payment.reference,
      },
    };
  });

  // Calculate the control sum (Just add up all payment amounts)
  let ctrlSum = parsedCreditTransfer.payments.reduce(
    (acc, payment) => acc + payment.amount,
    0
  );

  // round to 2 decimal places
  ctrlSum = Math.round(ctrlSum * 100) / 100;

  const doc = create({
    Document: {
      '@xmlns': 'urn:iso:std:iso:20022:tech:xsd:pain.001.001.03',
      '@xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
      '@xmlns:xsd': 'http://www.w3.org/2001/XMLSchema',
      CstmrCdtTrfInitn: {
        GrpHdr: {
          MsgId: parsedCreditTransfer.documentId,
          CreDtTm: parsedCreditTransfer.documentCreationDate
            .toISOString()
            .substring(0, 19), // Remove milliseconds and timezone
          NbOfTxs: parsedCreditTransfer.payments.length,
          CtrlSum: ctrlSum,
          InitgPty: {
            Nm: parsedCreditTransfer.initiatorName,
          },
        },
        PmtInf: {
          PmtInfId: parsedCreditTransfer.documentId,
          PmtMtd: 'TRF',
          BtchBookg: parsedCreditTransfer.batchBooking,
          NbOfTxs: parsedCreditTransfer.payments.length,
          CtrlSum: ctrlSum,
          ReqdExctnDt: parsedCreditTransfer.requestedExecutionDate
            .toISOString()
            .substring(0, 10), // Only keep the date
          Dbtr: {
            Nm: parsedCreditTransfer.initiatorName,
          },
          DbtrAcct: {
            Id: {
              IBAN: parsedCreditTransfer.initiatorIBAN,
            },
          },
          DbtrAgt: {
            FinInstnId: {
              BIC: parsedCreditTransfer.initiatorBIC,
            },
          },
          CdtTrfTxInf: mappedPayments,
        },
      },
    },
  });
  return [doc.end({ prettyPrint: true }), parsingResult.data];
};
