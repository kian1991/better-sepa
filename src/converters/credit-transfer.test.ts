import { createCreditTransferXML } from './credit-transfer';
import { CreditTransfer, CreditTransferInput } from '../types/types';

describe('createCreditTransferXML', () => {
	it('should create a valid XML document', () => {
		const creditTransfer: CreditTransferInput = {
			documentCreationDate: new Date('2021-01-01'),
			initiatorName: 'John Doe',
			initiatorIBAN: 'DE12100110012281061880',
			initiatorBIC: 'DEUTDEFF',
			requestedExecutionDate: new Date('2021-01-02'),
			batchBooking: false,
			payments: [
				{
					reference: 'Test',
					recipientIBAN: 'DE12100110012281061880',
					amount: 3434,
					recipientName: 'Jane Doe',
					end2endReference: '123456789',
				},
			],
		};

		const xml = createCreditTransferXML(creditTransfer);
		console.log(xml);
		expect(xml).toMatchSnapshot();
	});
});
