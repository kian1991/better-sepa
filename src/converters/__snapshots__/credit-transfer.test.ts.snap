// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCreditTransferXML should create a valid XML document 1`] = `
"<?xml version="1.0"?>
<Document xmlns="urn:iso:std:iso:20022:tech:xsd:pain.001.001.03" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:iso:std:iso:20022:tech:xsd:pain.001.001.03 pain.001.001.03.xsd">
  <CstmrCdtTrfInitn>
    <GrpHdr>
      <MsgId>PAIN0012024052309032251</MsgId>
      <CreDtTm>2021-01-01T00:00:00.000Z</CreDtTm>
      <NbOfTxs>1</NbOfTxs>
      <CtrlSum>3434</CtrlSum>
      <InitgPty>
        <Nm>John Doe</Nm>
      </InitgPty>
    </GrpHdr>
    <PmtInf>
      <PmtInfId>PAIN0012024052309032251</PmtInfId>
      <PmtMtd>TRF</PmtMtd>
      <BtchBookg>false</BtchBookg>
      <ReqdExctnDt>2021-01-02T00:00:00.000Z</ReqdExctnDt>
      <Dbtr>
        <Nm>John Doe</Nm>
      </Dbtr>
      <DbtrAcct>
        <Id>
          <IBAN>DE12100110012281061880</IBAN>
        </Id>
      </DbtrAcct>
      <DbtrAgt>
        <FinInstnId>
          <BIC>DEUTDEFF</BIC>
        </FinInstnId>
      </DbtrAgt>
      <CdtTrfTxInf>
        <PmtId>
          <EndToEndId>123456789</EndToEndId>
        </PmtId>
        <Amt>
          <InstdAmt Ccy="EUR">3434</InstdAmt>
        </Amt>
        <CdtrAgt>
          <FinInstnId/>
        </CdtrAgt>
        <Cdtr>
          <Nm>Jane Doe</Nm>
        </Cdtr>
        <CdtrAcct>
          <Id>
            <IBAN>DE12100110012281061880</IBAN>
          </Id>
          <Agt>
            <FinInstnId>
              <Othr>
                <Id>NOTPROVIDED</Id>
              </Othr>
            </FinInstnId>
          </Agt>
          <RmtInf>
            <Ustrd>Test</Ustrd>
          </RmtInf>
        </CdtrAcct>
      </CdtTrfTxInf>
    </PmtInf>
  </CstmrCdtTrfInitn>
</Document>"
`;
